server/service/src/vaccine_course/update.rs

```diff
@@ -0,0 +1,180 @@
+            id: self.id,
+            dose_number: self.dose_number,
+            label: self.label,
+            vaccine_course_id,
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Clone, Default)]
+pub struct UpdateVaccineCourse {
+    pub id: String,
+    pub name: Option<String>,
+    pub item_ids: Vec<VaccineCourseItem>,
+    pub schedules: Vec<VaccineCourseSchedule>,
+    pub demographic_indicator_id: Option<String>,
+    pub coverage_rate: f64,
+    pub is_active: bool,
+    pub wastage_rate: f64,
+    pub doses: i32,
+}
+
+pub fn update_vaccine_course(
+    ctx: &ServiceContext,
+    input: UpdateVaccineCourse,
+) -> Result<VaccineCourseRow, UpdateVaccineCourseError> {
+    let vaccine_course = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let old_row = validate(&input, connection)?;
+            let new_vaccine_course = generate(old_row, input.clone());
+            VaccineCourseRowRepository::new(connection).upsert_one(&new_vaccine_course)?;
+
+            // Update ITEMS - Delete and recreate all records.
+            // If nothing has changed, we still need to query and compare each record so this is the simplest way
+            let item_repo = VaccineCourseItemRowRepository::new(connection);
+            // Delete the existing vaccine course items
+            item_repo.delete_by_vaccine_course_id(&new_vaccine_course.id)?;
+
+            // Insert the new vaccine course items
+            for item in input.clone().item_ids {
+                item_repo.upsert_one(&item.to_domain(input.clone().id))?;
+            }
+
+            // Update Schedules - Delete and recreate all records.
+            let schedule_repo = VaccineCourseScheduleRowRepository::new(connection);
+            // Delete the existing vaccine course schedules
+            schedule_repo.delete_by_vaccine_course_id(&new_vaccine_course.id)?;
+
+            // Insert the new vaccine course schedules
+            for schedule in input.clone().schedules {
+                schedule_repo.upsert_one(&schedule.to_domain(input.clone().id))?;
+            }
+
+            activity_log_entry(
+                ctx,
+                ActivityLogType::VaccineCourseUpdated,
+                Some(new_vaccine_course.id.clone()),
+                None,
+                None,
+            )?;
+
+            get_vaccine_course(&ctx.connection, new_vaccine_course.id)
+                .map_err(UpdateVaccineCourseError::from)
+        })
+        .map_err(|error| error.to_inner_error())?;
+    Ok(vaccine_course)
+}
+
+pub fn validate(
+    input: &UpdateVaccineCourse,
+    connection: &StorageConnection,
+) -> Result<VaccineCourseRow, UpdateVaccineCourseError> {
+    let result = check_vaccine_course_exists(&input.id, connection)?;
+
+    let old_row = match result {
+        Some(vaccine_course) => vaccine_course,
+        None => return Err(UpdateVaccineCourseError::VaccineCourseDoesNotExist),
+    };
+
+    // TODO: demographic_indicator_id
```

Comment:
RE-activate this  code and add test?

Reply:
I'll do this one...

Reply:
[cd8fca5](https://github.com/msupply-foundation/open-msupply/pull/4142/commits/cd8fca52bcba15450f42d2e374abe23529748830)